File: src/data/o1/Sorted_GCD_Pair_Queries/Sorted_GCD_Pair_Queries.cpp
#include <vector>
#include <algorithm>
#include <numeric>

using namespace std;

class Solution {
public:
    vector<int> gcdValues(vector<int>& nums, vector<long long>& queries) {
        int n = nums.size();
        // Calculate all possible GCDs of unique pairs
        vector<int> gcdPairs;
        gcdPairs.reserve((n * (n - 1)) / 2);
        
        for(int i = 0; i < n; ++i){
            for(int j = i + 1; j < n; ++j){
                gcdPairs.emplace_back(__gcd(nums[i], nums[j]));
            }
        }
        
        // Sort the GCDs in ascending order
        sort(gcdPairs.begin(), gcdPairs.end());
        
        // Prepare the answer for each query
        vector<int> answer;
        answer.reserve(queries.size());
        for(auto q : queries){
            if(q >= 0 && q < gcdPairs.size()){
                answer.emplace_back(gcdPairs[q]);
            }
            else{
                answer.emplace_back(-1); // Assuming -1 for out-of-bound queries
            }
        }
        
        return answer;
    }
};

//  Test cases for the C++ solution
#include <cassert>

void testSolution(Solution& solution) {

    // Test case 1: Simple case with two numbers
    std::vector<int> nums = {2, 3};
    std::vector<long long> queries = {0};
    std::vector<int> expected_result = {1};  // Only one pair, GCD(2, 3) = 1
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 1 failed. Expected {1}.");

    // Test case 2: Multiple pairs with diverse GCDs
    nums = {2, 3, 4};
    queries = {0, 2, 2};
    expected_result = {1, 2, 2};  // GCD pairs sorted: [2, 2, 1]
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 2 failed. Expected {1, 2, 2}.");

    // Test case 3: All numbers are multiples of a common number
    nums = {4, 8, 12};
    queries = {0, 1, 2};
    expected_result = {4, 4, 4};  // GCD pairs sorted: [4, 4, 4]
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 3 failed. Expected {4, 4, 4}.");

    // Test case 4: GCD results in multiple distinct values
    nums = {2, 3, 5, 7};
    queries = {0, 1, 2, 3, 4, 5};
    expected_result = {1, 1, 1, 1, 1, 1};  // GCDs are all 1 for prime pairs
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 4 failed. Expected {1, 1, 1, 1, 1, 1}.");

    // Test case 5: GCD values with repeated elements in nums
    nums = {4, 4, 2, 1};
    queries = {5, 3, 1, 0};
    expected_result = {4, 2, 1, 1};  // GCD pairs sorted: [4, 4, 2, 2, 2, 1]
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 5 failed. Expected {4, 2, 1, 1}.");

    // Test case 6: Large number with single pair
    nums = {100000, 100000};
    queries = {0};
    expected_result = {100000};  // Only one GCD of the pair: [100000]
    assert(solution.gcdValues(nums, queries) == expected_result && "Test case 6 failed. Expected {100000}.");
}


int main() {
    Solution solution;
    for (int i = 0; i < 1280; ++i) {
        testSolution(solution);  
    }
    return 0;
}

File: src/data/o1/Sorted_GCD_Pair_Queries/Sorted_GCD_Pair_Queries.java
import java.util.*;

class Solution {
    public int[] gcdValues(int[] nums, long[] queries) {
        int n = nums.length;
        int maxNum = 0;
        for(int num : nums) {
            maxNum = Math.max(maxNum, num);
        }
        
        // Frequency of each number
        int[] freq = new int[maxNum +1];
        for(int num : nums) {
            freq[num]++;
        }
        
        // Compute frequency of multiples for each g
        long[] pairs = new long[maxNum +1];
        for(int g =1; g <= maxNum; g++) {
            int count =0;
            for(int multiple = g; multiple <= maxNum; multiple +=g) {
                count += freq[multiple];
            }
            pairs[g] = ((long)count * (count -1)) /2;
        }
        
        // Inclusion-Exclusion to get exact pairs with GCD =g
        for(int g = maxNum; g >=1; g--) {
            for(int multiple = 2*g; multiple <= maxNum; multiple +=g) {
                pairs[g] -= pairs[multiple];
            }
        }
        
        // Create list of GCDs in ascending order with their counts
        List<int[]> gcdList = new ArrayList<>();
        for(int g =1; g <= maxNum; g++) {
            if(pairs[g] >0){
                gcdList.add(new int[]{g, (int)pairs[g]});
            }
        }
        
        // Sort the gcdList by g in ascending order
        gcdList.sort(Comparator.comparingInt(a -> a[0]));
        
        // Create prefix sum
        long[] prefixSum = new long[gcdList.size()];
        long sum =0;
        for(int i =0; i < gcdList.size(); i++) {
            sum += gcdList.get(i)[1];
            prefixSum[i] = sum;
        }
        
        // Answer the queries
        int q = queries.length;
        int[] answer = new int[q];
        for(int i =0; i < q; i++) {
            long query = queries[i];
            if(query <0 || query >= prefixSum[prefixSum.length -1]){
                answer[i] = -1; // Or handle as per problem constraints
                continue;
            }
            // Binary search to find the smallest g where prefixSum[g] > query
            int left =0, right = prefixSum.length -1;
            int res = -1;
            while(left <= right){
                int mid = left + (right - left)/2;
                if(prefixSum[mid] > query){
                    res = mid;
                    right = mid -1;
                }
                else{
                    left = mid +1;
                }
            }
            if(res != -1){
                answer[i] = gcdList.get(res)[0];
            }
            else{
                answer[i] = -1;
            }
        }
        
        return answer;
    }
}

public class Sorted_GCD_Pair_Queries {
    public static void runTests(Solution solution) {
        // Test case 1
        int[] nums = {2, 3};
        long[] queries = {0};
        int[] expectedResult = {1};  // Only one pair, GCD(2, 3) = 1
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 1 failed. Expected " + java.util.Arrays.toString(expectedResult);

        // Test case 2
        nums = new int[]{2, 3, 4};
        queries = new long[]{0, 2, 2};
        expectedResult = new int[]{1, 2, 2};  // GCD pairs sorted: [2, 2, 1]
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 2 failed. Expected " + java.util.Arrays.toString(expectedResult);

        // Test case 3
        nums = new int[]{4, 8, 12};
        queries = new long[]{0, 1, 2};
        expectedResult = new int[]{4, 4, 4};  // GCD pairs sorted: [4, 4, 4]
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 3 failed. Expected " + java.util.Arrays.toString(expectedResult);

        // Test case 4
        nums = new int[]{2, 3, 5, 7};
        queries = new long[]{0, 1, 2, 3, 4, 5};
        expectedResult = new int[]{1, 1, 1, 1, 1, 1};  // GCDs are all 1 for prime pairs
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 4 failed. Expected " + java.util.Arrays.toString(expectedResult);

        // Test case 5
        nums = new int[]{4, 4, 2, 1};
        queries = new long[]{5, 3, 1, 0};
        expectedResult = new int[]{4, 2, 1, 1};  // GCD pairs sorted: [4, 4, 2, 2, 2, 1]
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 5 failed. Expected " + java.util.Arrays.toString(expectedResult);

        // Test case 6
        nums = new int[]{100000, 100000};
        queries = new long[]{0};
        expectedResult = new int[]{100000};  // Only one GCD of the pair: [100000]
        assert java.util.Arrays.equals(solution.gcdValues(nums, queries), expectedResult)
                : "Test case 6 failed. Expected " + java.util.Arrays.toString(expectedResult);
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 20000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    